src/consumer_queue.cpp

```diff
@@ -27,6 +26,7 @@ void consumer_queue::push_sample(const sample_p &sample) {
 		sample_p dummy;
 		buffer_.pop(dummy);
 	}
+	cv_.notify_one();
```

Comment:
It might be necessary to lock the mutex before calling `cv_.notify_one()`, see https://stackoverflow.com/q/2763714/73299 but also https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one

This will be a lot less performant than the original version without the locks so I'll push a benchmark template later.

Reply:
Are you sure? In the second link they explicitly state "The notifying thread does not need to hold the lock on the same mutex as the one held by the waiting thread(s); in fact doing so is a pessimization, since the notified thread would immediately block again, waiting for the notifying thread to release the lock."  Stackoverflow vs Stackoverflow https://stackoverflow.com/a/17102100 :D 

I believe the lock in the notifying thread is here sometimes to prevent race condition -- which is what #71 is about.

Anyway, I can run some more test with and without lock if you want, with the test code you pushed :)

Reply:
I think it's needed in the following case:

- consumer checks for a new sample, doesn't find one
- consumer acquires lock
- producer adds a sample, calls `cv_.notify()`
- consumer calls `cv_.wait()`

Normally this wouldn't be a problem as the next enqueue operation would also notify the consumer, but it would (only sometimes) add the inter sample time as latency